jenkins-dsl/jenkins_jira_integration.py

```diff
@@ -0,0 +1,352 @@
+    """
+    For the input build back as far as we have history, we want to cache the following (k/v):
+        NUMBER:             str number of this build
+        PREVIOUS_NUMBER     str build number of the previous build to walk to when walking the cache
+        failures:           set of test failures seen for this build (note: sets aren't serializable in JSON so we cast to list)
+
+        We use the build number and previous number as strings since they're going to get case to that in the JSON
+        ser/deser anyway, so best not to mix.
+    :return: testnu->
+    """
+    if not os.path.isdir('cache'):
+        os.mkdir('cache')
+
+    cached_data = {}
+    json_file = 'cache/' + branch
+    if os.path.exists(json_file):
+        with open(json_file, 'r', encoding='utf-8') as infile:
+            cached_data = json.load(infile)
+        for build in cached_data:
+            log('Loaded build: [' + str(build) + '] from cache')
+
+    while buildnum != NO_PREVIOUS_BUILD:
+        log('Processing cache for build_number ' + buildnum + ' on branch [' + branch + ']')
+        if buildnum in cached_data:
+            build_data = cached_data[buildnum]
+            log('   Already found data for ' + buildnum + ' cached locally. Continuing to ' + str(build_data[PREVIOUS_NUMBER]))
+            buildnum = str(build_data[PREVIOUS_NUMBER])
+        else:
+            print('   Did not find build_number: ' + buildnum + ' in the json cache. Populating...')
+            # Pull the data from the Jenkins server as we don't have it in our cache
+            build_data = BuildData(server, branch, buildnum)
+
+            # The data in our JSON cache is a pretty simple subset of what we have in BuildData for a full CI report
+            newcache_data = {}
+            newcache_data[NUMBER] = build_data.number
+            newcache_data[PREVIOUS_NUMBER] = build_data.previous_number
+            # Can't serialize sets to JSON
+            newcache_data['failures'] = list(build_data.test_failures)
+
+            log('   Got data for ' + str(newcache_data[NUMBER]) + '. Caching and setting previous build pointer to: ' + str(newcache_data[PREVIOUS_NUMBER]))
+            if 'failures' in newcache_data:
+                log('   build number: ' + str(newcache_data[NUMBER]) + ' with failure count: ' + str(len(newcache_data['failures'])))
+
+            # JSON comes back out as a string; need to follow suit here so the cache is all the same data type
+            cached_data[str(build_data.number)] = newcache_data
+
+            # cast to str should be redundant but it's coming out as an int. Not worth arguing with it.
+            buildnum = str(newcache_data[PREVIOUS_NUMBER])
+
+    log('   Hit limit of builds. Saving JSON.')
+    with open(json_file, 'w', encoding='utf-8') as outfile:
+        log('Updating the cache with recently queried results')
+        json.dump(cached_data, outfile)
```

Comment:
Great that the json is a small subset of the original data. Do we have an idea of its actual size?

We have consistent problems with diskspace on the jenkins agents, if this script is intended later to be run as part of jenkin's builds. The cache would be created on every jenkins agent, and the script can be called concurrently (as there is configured two executors on each agent).

Is there any value in centralising the cache on nightlies.a.o (to avoid 40 agents all pulling the original json api files)?

Reply:
All the trunk builds we have in history are roughly 50k I believe. So I think we should be fine on space.

Reply:
gzip compresses well, we can always gzip the file to shrink this further